fix: replace ofetch native fetch

alvarosabu opened the pull request:
- Removed the `ofetch` dependency from multiple files and replaced it with a `customFetch` utility for better error handling and flexibility.
- Updated API calls in login, pull languages, and user actions to use the new fetch method.
- Introduced a new ResponseError interface for improved error management.
- Added a getStoryblokUrl utility function to centralize URL construction based on region.
- Enhanced error handling in API error management to accommodate the new FetchError class.

## How to test this PR

Unit tests remained the same. They are all passing

## Other information

alvarosabu committed:
refactor: replace ofetch with customFetch and improve error handling

- Removed the ofetch dependency from multiple files and replaced it with a customFetch utility for better error handling and flexibility.
- Updated API calls in login, pull languages, and user actions to use the new fetch method.
- Introduced a new ResponseError interface for improved error management.
- Added a getStoryblokUrl utility function to centralize URL construction based on region.
- Enhanced error handling in API error management to accommodate the new FetchError class.

alvarosabu committed:
chore: update lock

edodusi commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,82 @@
+  try {
+    const headers = {
+      'Content-Type': 'application/json',
+      ...options.headers,
```
since it's needed in every call, why don't you put the `Authorization: $token` here too? It's overridable, but it spares you from manually adding it to every single `customFetch` call

alvarosabu replied:
It's not needed in every call. Check login actions.

edodusi replied:
@alvarosabu ok, anyways it could be nullable. If you think this doesn't simplify things, feel free to resolve this one

edodusi commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,82 @@
+          fetchOptions.body = options.body
+        }
+        catch {
+          fetchOptions.body = JSON.stringify(options.body)
```
why you try to `stringify` a `string` here?

edodusi commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,82 @@
+        data = await response.json()
+      }
+      catch {
+        // Ignore JSON parse errors
```
are you sure we can ignore all `.json()` errors here?

edodusi commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,82 @@
+      })
+    }
+
+    // Handle empty responses
```
is this comment correct?

edodusi commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,82 @@
+    if (contentType?.includes('application/json')) {
+      return await response.json()
+    }
+    return await response.text() as T
```
I don't get why if the response is `json` then you don't type with the generics but otherwise you return as plain `text` but typed

alvarosabu committed:
feat: simplified fetch wrapper and enhance type safety

- Updated `NetrcMachine` and `SessionState` interfaces to use `RegionCode` for improved type safety.
- Refactored login actions to handle optional chaining for better error handling.
- Enhanced API calls to utilize the new `customFetch` utility with improved type definitions.
- Added tests for `customFetch` to ensure robust error handling and response management.

alvarosabu commented:
Hi @edodusi I simplified the fetch method, there was some over-engineering there, I also added some unit tests to it and typing using generics

edodusi commented on src/commands/user/index.ts:
```diff
@@ -23,6 +23,7 @@ export const userCommand = program
     try {
       const { password, region } = state as NetrcMachine
       const { user } = await getUser(password, region)
+      console.log(user)
```
possible leftover console log?

alvarosabu committed:
chore: remove leftover console.log

edodusi commented:
@alvarosabu everything works now, just a little note: if I launch the `test` command (I'm correctly logged in) I get this:

```
 Test  Attempting a test...
 
Error

▲ error HTTP error! status: 404
```

Is this ok for this command? Or unrelated with this PR?

alvarosabu commented:
> @alvarosabu everything works now, just a little note: if I launch the `test` command (I'm correctly logged in) I get this:
> 
> ```
>  Test  Attempting a test...
>  
> Error
> 
> ▲ error HTTP error! status: 404
> ```
> 
> Is this ok for this command? Or unrelated with this PR?

You can safely ignore the test command everywhere. It gives you 404 because it's pointing to an unexisting space:
